File: frameworks/runtime-src/Classes/yayoi/Piece.cpp
#include "Piece.h"

using namespace std;

namespace yayoi {

    Piece::Piece()
    :_masterId(0)
    ,_individualStatus({{POWER, 0}, {ARMOR, 0}, {RESIST, 0}})
    ,_hp(MAX_HP)
    ,_team(RED)
    ,_position({0, 0})
    ,_king(false)
    ,_idInMatch(0) {
    }

    Piece::Piece(int masterId, const map<Param, int> &individualStatus, Team team, int row, int col, bool king)
    :_masterId(masterId)
    ,_individualStatus(individualStatus)
    ,_hp(MAX_HP)
    ,_team(team)
    ,_position({row, col})
    ,_king(king)
    ,_idInMatch(0) {
    }

    Piece::~Piece() {
    }

    void Piece::setMaster(int masterId, Piece::Job job, Piece::Color color, const map<Param, int> &baseStatus) {
        Master master;
        master._job = job;
        master._color = color;
        master._baseStatus = baseStatus;
        getMasters()[masterId] = master;
    }

    map<int, Piece::Master>& Piece::getMasters() {
        static map<int, Piece::Master> masters;
        return masters;
    }

    void Piece::applyChip(const pair<int, int> &chip) {
        _position.first += chip.first;
        _position.second += chip.second;
    }

    int Piece::applyDamage(int damage) {
        damage = min(damage, _hp);
        _hp -= damage;
        if (_hp <= 0) {
            _position = {-1, -1};
        }
        return damage;
    }

}

File: frameworks/runtime-src/Classes/api/Piece.lua

--------------------------------
-- @module Piece
-- @parent_module yayoi

--------------------------------
-- 
-- @function [parent=#Piece] getColor 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] getStatus 
-- @param self
-- @param #int param
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] applyDamage 
-- @param self
-- @param #int damage
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] setIdInMatch 
-- @param self
-- @param #int idInMatch
        
--------------------------------
-- 
-- @function [parent=#Piece] getTeam 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] getJob 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] getBaseStatus 
-- @param self
-- @param #int param
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] applyChip 
-- @param self
-- @param #std::pair<int, int> chip
        
--------------------------------
-- 
-- @function [parent=#Piece] getMasterId 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] getIndividualStatus 
-- @param self
-- @param #int param
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] getHp 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] isKing 
-- @param self
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- 
-- @function [parent=#Piece] getIdInMatch 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#Piece] getPosition 
-- @param self
-- @return std::pair<int, int>#std::pair<int, int> ret (return value: std::pair<int, int>)
        
--------------------------------
-- 
-- @function [parent=#Piece] setMaster 
-- @param self
-- @param #int masterId
-- @param #int job
-- @param #int color
-- @param #map_table baseStatus
        
--------------------------------
-- @overload self, int, map_table, int, int, int, bool         
-- @overload self         
-- @function [parent=#Piece] Piece
-- @param self
-- @param #int masterId
-- @param #map_table individualStatus
-- @param #int team
-- @param #int row
-- @param #int col
-- @param #bool king

return nil
